consensus/blake3pow/poem.go

```diff
@@ -29,35 +29,45 @@ func (blake3pow *Blake3pow) CalcOrder(header *types.Header) (*big.Int, int, erro
+	// /////////////
+	// Compute the prime comulative entropy threshold based on the time factor of the hierarchy.
 
-	primeBlockEntropyThresholdAdder, _ := mathutil.BinaryLog(primeBlockThreshold, 8)
-	primeBlockEntropyThreshold := new(big.Int).Add(zoneThresholdS, big.NewInt(int64(primeBlockEntropyThresholdAdder)))
+	primeEntropyThresholdFactor := new(big.Int).Mul(timeFactorHierarchyDepthMultiple, timeFactorHierarchyDepthMultiple)
+	primeRandFactor := new(big.Int).Quo(primeEntropyThresholdFactor, big.NewInt(2))
+	primeRandFactorBits, _ := mathutil.BinaryLog(primeRandFactor, mantBits)
+	primeEntropyThresholdFactor = new(big.Int).Sub(primeEntropyThresholdFactor, big.NewInt(header.TerminusHash().GetLastNumBitsFromHash(primeRandFactorBits)))
+	primeEntropyThreshold := new(big.Int).Mul(primeEntropyThresholdFactor, zoneThresholdS)
 
-	totalDeltaS := new(big.Int).Add(header.ParentDeltaS(common.REGION_CTX), header.ParentDeltaS(common.ZONE_CTX))
-	totalDeltaS.Add(totalDeltaS, intrinsicS)
-	if intrinsicS.Cmp(primeBlockEntropyThreshold) > 0 && totalDeltaS.Cmp(primeEntropyThreshold) > 0 {
+	// Compute the total accumulated entropy since the last prime block
+	totalDeltaSPrime := new(big.Int).Add(header.ParentDeltaS(common.REGION_CTX), header.ParentDeltaS(common.ZONE_CTX))
+	totalDeltaSPrime.Add(totalDeltaSPrime, intrinsicS)
+
+	if totalDeltaSPrime.Cmp(primeEntropyThreshold) > 0 {
 		return intrinsicS, common.PRIME_CTX, nil
 	}
 
+	// ////////////
 	// Region case
-	regionEntropyThreshold := new(big.Int).Mul(timeFactorHierarchyDepthMultiple, zoneThresholdS)
-	regionBlockThreshold := new(big.Int).Quo(regionEntropyThreshold, big.NewInt(2))
-	regionEntropyThreshold = new(big.Int).Sub(regionEntropyThreshold, regionBlockThreshold)
-
-	regionBlockEntropyThresholdAdder, _ := mathutil.BinaryLog(regionBlockThreshold, 8)
-	regionBlockEntropyThreshold := new(big.Int).Add(zoneThresholdS, big.NewInt(int64(regionBlockEntropyThresholdAdder)))
-
-	totalDeltaS = new(big.Int).Add(header.ParentDeltaS(common.ZONE_CTX), intrinsicS)
-	if intrinsicS.Cmp(regionBlockEntropyThreshold) > 0 && totalDeltaS.Cmp(regionEntropyThreshold) > 0 {
+	// ////////////
+	// Compute the region comulative entropy threshold based on the time factor of the hierarchy.
+	regionEntropyThresholdFactor := timeFactorHierarchyDepthMultiple
+	regionRandFactor := new(big.Int).Quo(regionEntropyThresholdFactor, big.NewInt(2))
+	regionRandFactorBits, _ := mathutil.BinaryLog(regionRandFactor, mantBits)
+	regionEntropyThresholdFactor = new(big.Int).Sub(regionEntropyThresholdFactor, big.NewInt(header.TerminusHash().GetLastNumBitsFromHash(regionRandFactorBits)))
+	regionEntropyThreshold := new(big.Int).Mul(regionEntropyThresholdFactor, zoneThresholdS)
+
+	// Compute the total accumulated entropy since the last region block
+	totalDeltaSRegion := new(big.Int).Add(header.ParentDeltaS(common.ZONE_CTX), intrinsicS)
+
+	if totalDeltaSRegion.Cmp(regionEntropyThreshold) > 0 {
```

Comment:
I think I left this comment on the previous PR, but can't be sure.
1) it would be better to check the zone case first since it is the more frequent case, and no need to run through all the prime and region bigint math on _every_ zone block
2) I would strongly prefer these three blocks collapse into one, with a loop/index to cycle through the 1~2 parameters that makes each block unique.

Reply:
1. You can't logically invert these because you have to know the range.  If you find something that is a zone block you don't know it isn't a region or prime block without first doing the check on the region and prime block thresholds.

2. Tere is no way to collapse these loops.  The thing you could do is precompute the first three lines of each case during init.